src/apps/dialogs/TraitSelect/TraitSelectDialog.svelte

```diff
@@ -0,0 +1,225 @@
+            }
+        }
+
+        return true;
+    }
+
+    function isDisabled(trait) {
+        // Disable all nonchecked, check marks if we've reached selectionLimit
+        if (
+            selectionLimit &&
+            selected.length >= selectionLimit &&
+            !selected.includes(trait)
+        ) {
+            return true;
+        }
+
+        return disabled.includes(trait);
+    }
+
+    function setTrait(trait, status) {
+        if (status) {
+            selected = [...selected, trait];
+        } else {
+            selected.findSplice((selected) => selected === trait);
+        }
+
+        // Prevent duplicate values
+        selected = [...new Set(selected)];
+        updateDisabledStatus();
+    }
+
+    function getCustom() {
+        const custom = selected.filter(
+            (trait) => isCustomTrait(trait) && !_disabledCustom.includes(trait)
+        );
+
+        return custom.join(";");
+    }
+
+    function setCustom(element) {
+        // Split, trim, remove empty & disabled
+        const customList = element.value
+            .split(";")
+            .map((i) => i.trim())
+            .filter((i) => i && !_disabledCustom.includes(i));
+
+        selected = selected
+            // Remove all custom from selected
+            .filter((trait) => !isCustomTrait(trait))
+            // Add back disabled selected (our input box doesn't control these)
+            .concat(_disabledSelected);
+
+        // Test to ensure that the custom added selection isn't going to exceed the limit.
+        if (selectionLimit) {
+            // Calculate how much we're over the limit by.
+            const overLimit =
+                selected.length + customList.length - selectionLimit;
+
+            // Warn the user and truncate the customList
+            if (overLimit > 0) {
+                ui.notifications.warn(
+                    "You have added too many custom traits, we will remove the extra for you"
+                );
+                customList.splice(-overLimit);
+            }
+        }
+
+        // Add the custom traits to the selected and update
+        selected = [...selected, ...customList];
+        element.value = getCustom();
+        updateDisabledStatus();
+    }
+
+    function submitForm() {
+        application.submit(selected);
+    }
+
+    // Updates the disabled/checked values. Since any checkbox can modifiy the disabled state of any
+    // Due to the selectionLimit, we must maintain the list of _disabled elemets seperately so that
+    // Svelte knows to update those elements.
+    let _disabled = {};
+    let _disableSubmit = false;
+    function updateDisabledStatus() {
+        for (let group of traitGroups) {
+            for (let trait in group.traits ?? {}) {
+                _disabled[trait] = isDisabled(trait);
+            }
+        }
+
+        _disableSubmit = selectionLimit && selected.length !== selectionLimit;
+    }
```

Comment:
Same deal here with `let` vs `const`. Also, please make sure that function blocks are separated from other code.